addons/account_edi/models/account_edi_format.py

```diff
@@ -506,21 +506,90 @@ def _retrieve_partner(self, name=None, phone=None, mail=None, vat=None):
         :param vat:     The vat number of the partner.
         :returns:       A partner or an empty recordset if not found.
         '''
-        domains = []
-        for value, domain in (
-            (name, [('name', 'ilike', name)]),
-            (phone, expression.OR([[('phone', '=', phone)], [('mobile', '=', phone)]])),
-            (mail, [('email', '=', mail)]),
-            (vat, [('vat', 'like', vat)]),
-        ):
-            if value is not None:
-                domains.append(domain)
-
-        domain = expression.AND([
-            expression.OR(domains),
-            [('company_id', 'in', [False, self.env.company.id])],
-        ])
-        return self.env['res.partner'].search(domain, limit=1)
+        def search_with_vat(extra_domain):
+            if not vat:
+                return None
+
+            # Sometimes, the vat is specified with some whitespaces.
+            normalized_vat = vat.replace(' ', '')
+
+            partner = self.env['res.partner'].search([('vat', '=', normalized_vat)], limit=1)
+
+            # Try to remove the country code prefix from the vat.
+            if not partner and len(normalized_vat) > 2 and normalized_vat[:2].isalpha():
+                country_prefix = normalized_vat[:2]
+                normalized_vat = normalized_vat[2:]
```

Comment:
It is wrong to only consider the number without the country code if it starts with letters. Some countries use letters as part of the code (nothing related to country codes) and we store the country code in the database anyway

Reply:
This code is not perfect but managing all VAT perfectly implies to know all pattern in all countries. That's really complicated. It means comparing the VAT with all patterns to find the correct template, find a partner for each of them if the user didn't set a country_id on its partner

Reply:
my 2 cents, it won't work for NL. (example: NL123456789B01)

Reply:
Considering the first letters as being the country is wrong, see for instance. `UA` can
 be used in Mexican numbers.
https://github.com/odoo/odoo/blob/68cdf057a9e769071fab6726eaf3b37d16959bf1/addons/base_vat/models/res_partner.py#L481-L484